test/clj_kondo/main_test.clj

```diff
@@ -2483,6 +2483,32 @@
   (is (empty? (lint! "(ns main.core (:require [\"/vendors/daterangepicker\"]))"
                      "--lang" "cljc" "--cache" "true"))))
 
+(deftest lint-bangs!
+  (are [input expected] (testing input
```

Comment:
I know this `are` looks kind of funny but everything is there for a reason (`testing`, `pr-str`, `true` at tail position...)

The result being: tests that fail very informatively (the day they fail).

Also I tried to write the test in such a way that unrelated bugs don't make it fail.

I can refactor into sth else but also would be happy to explain a given detail.

Reply:
I broke the tests while trying to refactor and got this for output:

```
FAIL in (lint-bangs!) (main_test.clj:2487)
(defn a [] (((((((([[[1 1 1 1 (b!)]]])))))))))
({:file "<stdin>", :row 1, :col 32, :level :error, :message "unresolved symbol b!"})
expected: (= :invalid (get {0 :valid} (->> result (filter (comp #{"Functions using !-suffixed code should also be !-suffixed"} :message)) (count)) :invalid))
  actual: (not (= :invalid :valid))
```

I didn't find that very clear. I'd prefer if you refactored this into the same style of using `assert-submaps` like the other tests. That is also going to test the reported location, which is pretty important for in-editor usage of clj-kondo, which is a primary use case.

Reply:
I made a start with this in my PR.

Reply:
> I didn't find that very clear. 

`(not (= :invalid :valid))` is telling you `not= <expected> <actual>` in that specific order (as per usual in clojure.test)

Since the expectation was `:invalid` but the test case is `:valid`, something is wrong. `({:file "<stdin>", :row 1, :col 32, :level :error, :message "unresolved symbol b!"})` tells you the full linting output, so you can see that the output , in fact, lacks the precise message we wanted.

---

The big selling point of `are` is that you can have large, aligned, DRY rows of assertions in the following style:

```clj
 "(defn a [] [[[[1 1 1 (apply b!)]]]])" :invalid
```

It reads very similarly to a clojure hashmap `{input expectation}`.

As a result one can be encouraged to add _many_ test cases - more than people usually do IME.

Anyway, I'd be OK if we refactor prior to merging.